<a href="https://colab.research.google.com/github/IXIena/data-circle-team-a/blob/main/teamA_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [20]:
import io
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import mpl_toolkits
%matplotlib inline
from google.colab import drive
from google.colab import files
pd.set_option("display.max_columns", None)

In [21]:
from google.colab import drive
drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [22]:
df=pd.read_csv("/content/drive/MyDrive/Colab Notebooks/survey_results_public.csv")

# Exploration/Wrangling

In [33]:
df.iloc[:, :]

ResponseId     Q120                      MainBranch  \
0               1  I agree                   None of these   
1               2  I agree  I am a developer by profession   
2               3  I agree  I am a developer by profession   
3               4  I agree  I am a developer by profession   
4               5  I agree  I am a developer by profession   
...           ...      ...                             ...   
89179       89180  I agree  I am a developer by profession   
89180       89181  I agree  I am a developer by profession   
89181       89182  I agree     I code primarily as a hobby   
89182       89183  I agree  I am a developer by profession   
89183       89184  I agree  I am a developer by profession   

                      Age                                         Employment  \
0         18-24 years old                                                NaN   
1         25-34 years old                                Employed, full-time   
2         45-54 years old                                Employed, full-time   
3         25-34 years old                                Employed, full-time   
4         25-34 years old  Employed, full-time;Independent contractor, fr...   
...                   ...                                                ...   
89179     25-34 years old  Employed, full-time;Independent contractor, fr...   
89180     18-24 years old             Student, full-time;Employed, part-time   
89181   Prefer not to say                                I prefer not to say   
89182  Under 18 years old             Employed, part-time;Student, part-time   
89183     35-44 years old                                Employed, full-time   

                                 RemoteWork  \
0                                       NaN   
1                                    Remote   
2      Hybrid (some remote, some in-person)   
3      Hybrid (some remote, some in-person)   
4                                    Remote   
...                                     ...   
89179                                Remote   
89180  Hybrid (some remote, some in-person)   
89181                                   NaN   
89182  Hybrid (some remote, some in-person)   
89183  Hybrid (some remote, some in-person)   

                                        CodingActivities  \
0                                                    NaN   
1      Hobby;Contribute to open-source projects;Boots...   
2      Hobby;Professional development or self-paced l...   
3                                                  Hobby   
4      Hobby;Contribute to open-source projects;Profe...   
...                                                  ...   
89179  Hobby;Bootstrapping a business;Freelance/contr...   
89180                            School or academic work   
89181                                                NaN   
89182                      Hobby;School or academic work   
89183  Hobby;Professional development or self-paced l...   

                                                 EdLevel  \
0                                                    NaN   
1           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
...                                                  ...   
89179                Associate degree (A.A., A.S., etc.)   
89180       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
89181                                     Something else   
89182  Secondary school (e.g. American high school, G...   
89183       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                               LearnCode  \
0                                                    NaN   
1      Books / Physical media;Colleague;Friend or fam...   
2      Books / Physical media;Colleague;On the job tr...   
3      Colleague;Friend or family member;Other online.

In [ ]:
df.head(5)

ResponseId     Q120                      MainBranch              Age  \
0           1  I agree                   None of these  18-24 years old   
1           2  I agree  I am a developer by profession  25-34 years old   
2           3  I agree  I am a developer by profession  45-54 years old   
3           4  I agree  I am a developer by profession  25-34 years old   
4           5  I agree  I am a developer by profession  25-34 years old   

                                          Employment  \
0                                                NaN   
1                                Employed, full-time   
2                                Employed, full-time   
3                                Employed, full-time   
4  Employed, full-time;Independent contractor, fr...   

                             RemoteWork  \
0                                   NaN   
1                                Remote   
2  Hybrid (some remote, some in-person)   
3  Hybrid (some remote, some in-person)   
4                                Remote   

                                    CodingActivities  \
0                                                NaN   
1  Hobby;Contribute to open-source projects;Boots...   
2  Hobby;Professional development or self-paced l...   
3                                              Hobby   
4  Hobby;Contribute to open-source projects;Profe...   

                                        EdLevel  \
0                                           NaN   
1  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                           LearnCode  \
0                                                NaN   
1  Books / Physical media;Colleague;Friend or fam...   
2  Books / Physical media;Colleague;On the job tr...   
3  Colleague;Friend or family member;Other online...   
4  Books / Physical media;Online Courses or Certi...   

                                     LearnCodeOnline  LearnCodeCoursesCert  \
0                                                NaN                   NaN   
1  Formal documentation provided by the owner of ...                 Other   
2  Formal documentation provided by the owner of ...                   NaN   
3  Formal documentation provided by the owner of ...                   NaN   
4  Formal documentation provided by the owner of ...  Other;Codecademy;edX   

  YearsCode YearsCodePro                               DevType  \
0       NaN          NaN                                   NaN   
1        18            9  Senior Executive (C-Suite, VP, etc.)   
2        27           23                   Developer, back-end   
3        12            7                  Developer, front-end   
4         6            4                 Developer, full-stack   

                    OrgSize                 PurchaseInfluence      TechList  \
0                       NaN                               NaN           NaN   
1          2 to 9 employees  I have a great deal of influence   Investigate   
2  5,000 to 9,999 employees             I have some influence  Given a list   
3      100 to 499 employees             I have some influence   Investigate   
4        20 to 99 employees             I have some influence   Investigate   

                                          BuyNewTool  \
0                                                NaN   
1  Start a free trial;Ask developers I know/work ...   
2  Start a free trial;Ask developers I know/work ...   
3  Start a free trial;Ask developers I know/work ...   
4  Start a free trial;Ask developers I know/work ...   

                    Country                   Currency  CompTotal  \
0                       NaN                        NaN        NaN   
1  United States of America  USD\tUnited States dollar   285000.0   
2  United States of America  USD\tUnited States dollar   250000.0   
3  United States of A

In [34]:
# Check if the original DataFrame has duplicates
if df.duplicated().any():
    print("Original DataFrame has duplicates.")
else:
    print("Original DataFrame does not have duplicates.")


Original DataFrame does not have duplicates.


In [35]:
print("Number of duplicates in original DataFrame:", df.duplicated().sum())

Number of duplicates in original DataFrame: 0


In [36]:
df.describe(include='all')

ResponseId     Q120                      MainBranch  \
count   89184.000000    89184                           89184   
unique           NaN        1                               6   
top              NaN  I agree  I am a developer by profession   
freq             NaN    89184                           67237   
mean    44592.500000      NaN                             NaN   
std     25745.347541      NaN                             NaN   
min         1.000000      NaN                             NaN   
25%     22296.750000      NaN                             NaN   
50%     44592.500000      NaN                             NaN   
75%     66888.250000      NaN                             NaN   
max     89184.000000      NaN                             NaN   

                    Age           Employment  \
count             89184                87898   
unique                8                  106   
top     25-34 years old  Employed, full-time   
freq              33247                53748   
mean                NaN                  NaN   
std                 NaN                  NaN   
min                 NaN                  NaN   
25%                 NaN                  NaN   
50%                 NaN                  NaN   
75%                 NaN                  NaN   
max                 NaN                  NaN   

                                  RemoteWork CodingActivities  \
count                                  73810            73764   
unique                                     3              116   
top     Hybrid (some remote, some in-person)            Hobby   
freq                                   31131            15308   
mean                                     NaN              NaN   
std                                      NaN              NaN   
min                                      NaN              NaN   
25%                                      NaN              NaN   
50%                                      NaN              NaN   
75%                                      NaN              NaN   
max                                      NaN              NaN   

                                             EdLevel  \
count                                          87973   
unique                                             8   
top     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
freq                                           36706   
mean                                             NaN   
std                                              NaN   
min                                              NaN   
25%                                              NaN   
50%                                              NaN   
75%                                              NaN   
max                                              NaN   

                                                LearnCode  \
count                                               87663   
unique                                                790   
top     Other online resources (e.g., videos, blogs, f...   
freq                                                 4873   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                          LearnCodeOnline  \
count                                               70084   
unique                                               7940   
top     Formal documentation provided by the owner of ...   
freq                                                 1182   
mean                                                  NaN   
std                                                   NaN   
min                                   

In [37]:
df.describe()

ResponseId     CompTotal       WorkExp  ConvertedCompYearly
count  89184.000000  4.822500e+04  43579.000000         4.801900e+04
mean   44592.500000  1.036807e+42     11.405126         1.031101e+05
std    25745.347541  2.276847e+44      9.051989         6.814188e+05
min        1.000000  0.000000e+00      0.000000         1.000000e+00
25%    22296.750000  6.300000e+04      5.000000         4.390700e+04
50%    44592.500000  1.150000e+05      9.000000         7.496300e+04
75%    66888.250000  2.300000e+05     16.000000         1.216410e+05
max    89184.000000  5.000000e+46     50.000000         7.435143e+07

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89184 entries, 0 to 89183
Data columns (total 84 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   ResponseId                           89184 non-null  int64  
 1   Q120                                 89184 non-null  object 
 2   MainBranch                           89184 non-null  object 
 3   Age                                  89184 non-null  object 
 4   Employment                           87898 non-null  object 
 5   RemoteWork                           73810 non-null  object 
 6   CodingActivities                     73764 non-null  object 
 7   EdLevel                              87973 non-null  object 
 8   LearnCode                            87663 non-null  object 
 9   LearnCodeOnline                      70084 non-null  object 
 10  LearnCodeCoursesCert                 37076 non-null  object 
 11  YearsCode                   

In [39]:
df.isna().any()
df.isna().sum()

ResponseId                 0
Q120                       0
MainBranch                 0
Age                        0
Employment              1286
                       ...  
ProfessionalTech       47401
Industry               52410
SurveyLength            2699
SurveyEase              2630
ConvertedCompYearly    41165
Length: 84, dtype: int64

In [40]:
df_null_percent = (df.isna().sum()*100 / df['ResponseId'].count())

In [41]:
df_null_percent

ResponseId              0.000000
Q120                    0.000000
MainBranch              0.000000
Age                     0.000000
Employment              1.441963
                         ...    
ProfessionalTech       53.149668
Industry               58.766146
SurveyLength            3.026328
SurveyEase              2.948959
ConvertedCompYearly    46.157382
Length: 84, dtype: float64

# Cleaning

In [42]:
sorted_null_percentages = df_null_percent.sort_values(ascending=False)

In [43]:
columns_above_50_percent_null = df_null_percent[df_null_percent >= 50]

print(columns_above_50_percent_null)

LearnCodeCoursesCert                   58.427521
AIDevHaveWorkedWith                    70.954431
AIDevWantToWorkWith                    78.037540
AIAcc                                  56.725422
AIToolInterested in Using              63.241164
AIToolCurrently Using                  59.480400
AIToolNot interested in Using          76.375807
AINextVery different                   85.803507
AINextNeither different nor similar    92.600691
AINextSomewhat similar                 93.005472
AINextVery similar                     97.061132
AINextSomewhat different               73.870874
ICorPM                                 51.036060
WorkExp                                51.135854
Knowledge_1                            52.306468
Knowledge_2                            53.276372
Knowledge_3                            53.132849
Knowledge_4                            53.260675
Knowledge_5                            53.436715
Knowledge_6                            53.444564
Knowledge_7         

In [44]:
columns_above_50_percent_null = sorted_null_percentages[sorted_null_percentages >= 50]

print(columns_above_50_percent_null)

AINextVery similar                     97.061132
AINextSomewhat similar                 93.005472
AINextNeither different nor similar    92.600691
AINextVery different                   85.803507
AIDevWantToWorkWith                    78.037540
AIToolNot interested in Using          76.375807
AINextSomewhat different               73.870874
AIDevHaveWorkedWith                    70.954431
AIToolInterested in Using              63.241164
AIToolCurrently Using                  59.480400
Industry                               58.766146
LearnCodeCoursesCert                   58.427521
AIAcc                                  56.725422
Frequency_3                            53.967079
Knowledge_8                            53.574632
Knowledge_7                            53.503992
Knowledge_6                            53.444564
Knowledge_5                            53.436715
Knowledge_2                            53.276372
Knowledge_4                            53.260675
ProfessionalTech    

In [45]:
columns_above_50_percent_null = sorted_null_percentages[sorted_null_percentages >= 50]

In [46]:
df_reduced = df.drop(columns=columns_above_50_percent_null.index, axis=1)

In [47]:
df_reduced.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89184 entries, 0 to 89183
Data columns (total 55 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   ResponseId                      89184 non-null  int64  
 1   Q120                            89184 non-null  object 
 2   MainBranch                      89184 non-null  object 
 3   Age                             89184 non-null  object 
 4   Employment                      87898 non-null  object 
 5   RemoteWork                      73810 non-null  object 
 6   CodingActivities                73764 non-null  object 
 7   EdLevel                         87973 non-null  object 
 8   LearnCode                       87663 non-null  object 
 9   LearnCodeOnline                 70084 non-null  object 
 10  YearsCode                       87435 non-null  object 
 11  YearsCodePro                    66136 non-null  object 
 12  DevType                         

In [48]:
import re
column_name = 'Age'

# Use regular expressions to keep only numeric characters and hyphens
df_reduced[column_name] = df_reduced[column_name].astype(str).apply(lambda x: re.sub(r'[^0-9-]', '', x))

In [49]:
column_contents = df_reduced['Age']

# Display the contents of the column
print(column_contents)

0        18-24
1        25-34
2        45-54
3        25-34
4        25-34
         ...  
89179    25-34
89180    18-24
89181         
89182       18
89183    35-44
Name: Age, Length: 89184, dtype: object


In [50]:
df_new = df_reduced.copy()

# Identifying Salary Column


The first set of statistics presents mean and median values within a reasonable range for salary data, with mean around 103,000 and median around 75,000. In contrast, the second set exhibits significantly higher mean and median values, with the mean being approximately 10^42 and the median being in the order of 10^5.

Moreover, the standard deviation in the first set is relatively high (681,418), indicative of expected variability in salary data. Conversely, the standard deviation in the second set is astronomically high (2.28 x 10^44), suggesting atypical dispersion not consistent with salary distributions.

Regarding the range of values, the "ConvertedCompYearly" column contains figures within a plausible salary range, and its frequency distribution aligns with expectations for salary data. Conversely, the "CompTotal" column includes many values with exceptionally high magnitudes, such as 3.3e+09, which are unlikely to represent individual salaries and may indicate total compensation or other financial metrics.

Considering these observations alongside the column names, "ConvertedCompYearly" appears more likely to represent salary data, given its alignment with salary-related terminology and distribution characteristics.

In [28]:
possible_salary_column = df_reduced['ConvertedCompYearly']

print("Basic statistics:")
print(possible_salary_column.describe())

Basic statistics:
count    4.801900e+04
mean     1.031101e+05
std      6.814188e+05
min      1.000000e+00
25%      4.390700e+04
50%      7.496300e+04
75%      1.216410e+05
max      7.435143e+07
Name: ConvertedCompYearly, dtype: float64


In [32]:
possible_salary_column2 =  df_reduced['CompTotal']

print("Basic statistics:")
print(possible_salary_column2.describe())

Basic statistics:
count    4.822500e+04
mean     1.036807e+42
std      2.276847e+44
min      0.000000e+00
25%      6.300000e+04
50%      1.150000e+05
75%      2.300000e+05
max      5.000000e+46
Name: CompTotal, dtype: float64


In [29]:
df_reduced.ConvertedCompYearly.value_counts()

ConvertedCompYearly
64254.0     784
53545.0     615
150000.0    585
200000.0    558
74963.0     557
           ... 
31494.0       1
859.0         1
77904.0       1
74917.0       1
16917.0       1
Name: count, Length: 8784, dtype: int64

In [30]:
df_reduced.CompTotal.value_counts()

CompTotal
1.000000e+05    1355
6.000000e+04    1290
1.200000e+05    1240
1.500000e+05    1150
8.000000e+04    1066
                ... 
1.642800e+06       1
1.380910e+05       1
2.957000e+04       1
4.190000e+05       1
3.300000e+09       1
Name: count, Length: 3828, dtype: int64

# Nowa sekcja